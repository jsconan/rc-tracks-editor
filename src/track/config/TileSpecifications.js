/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * Gives the specifications for the track tiles with respect to the initial constraints.
 */
export class TileSpecifications {
    /**
     * Defines the specifications for the track tiles from the given constraints.
     * @param {number} laneWidth - The width of the track lane (the distance between the barriers).
     * @param {number} barrierWidth - The width of the barriers.
     * @param {number} barrierChunks - The number of barrier chunks per tile.
     * @param {number} maxRatio - The maximum value for size ratios.
     */
    constructor(laneWidth = 20, barrierWidth = 1, barrierChunks = 4, maxRatio = 4) {
        this.setLaneWidth(laneWidth);
        this.setBarrierWidth(barrierWidth);
        this.setBarrierChunks(barrierChunks);
        this.setMaxRatio(maxRatio);
    }

    /**
     * The length of a tile with respect to the initial constraints.
     * @type {number}
     */
    get length() {
        return this.laneWidth * 1.25 + this.barrierWidth * 2;
    }

    /**
     * The width of a tile with respect to the initial constraints.
     * @type {number}
     */
    get width() {
        return this.laneWidth + this.barrierWidth * 2;
    }

    /**
     * The length of a barrier chunk with respect to the initial constraints.
     * @type {number}
     */
    get barrierLength() {
        return this.length / this.barrierChunks;
    }

    /**
     * The padding around the width of the tile to align with its length.
     * Note: tiles must fit in a square so that they can be flipped and rotated indifferently.
     * @type {number}
     */
    get padding() {
        return this.laneWidth / 8;
    }

    /**
     * Sets the width of the track lane (the distance between the barriers).
     * @param {number} laneWidth - The width of the track lane.
     * @returns {TileSpecifications} - Chains the instance.
     */
    setLaneWidth(laneWidth) {
        this.laneWidth = Math.abs(laneWidth);

        return this;
    }

    /**
     * Sets the width of the barriers.
     * @param {number} barrierWidth - The width of the barriers.
     * @returns {TileSpecifications} - Chains the instance.
     */
    setBarrierWidth(barrierWidth) {
        this.barrierWidth = Math.abs(barrierWidth);

        return this;
    }

    /**
     * Sets the number of barrier chunks per tile.
     * @param {number} barrierChunks - The number of barrier chunks per tile.
     * @returns {TileSpecifications} - Chains the instance.
     */
    setBarrierChunks(barrierChunks) {
        this.barrierChunks = Math.abs(Math.round(barrierChunks) || 1);

        return this;
    }

    /**
     * Sets the maximum value for size ratios.
     * @param {number} maxRatio - The maximum value for size ratios.
     * @returns {TileSpecifications} - Chains the instance.
     */
    setMaxRatio(maxRatio) {
        this.maxRatio = Math.abs(Math.round(maxRatio) || 1);

        return this;
    }
}
