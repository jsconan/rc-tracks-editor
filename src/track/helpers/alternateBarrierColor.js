/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { alternate } from '../../core/helpers';

/**
 * A list of colors for the barriers.
 * @type {string[]}
 */
const barrierColors = ['even', 'odd'];

/**
 * Returns the color for a barrier element at the given position.
 * @param {number} i - The position of the barrier element.
 * @returns {string} - The color for the barrier element.
 */
export default i => alternate(i, barrierColors);
