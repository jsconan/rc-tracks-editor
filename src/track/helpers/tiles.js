/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * Tile oriented to the right.
 * @type {string}
 */
export const TILE_DIRECTION_RIGHT = 'right';

/**
 * Tile oriented to the left.
 * @type {string}
 */
export const TILE_DIRECTION_LEFT = 'left';

/**
 * Type for a default tile.
 * @type {string}
 */
export const DEFAULT_TILE_TYPE = 'tile';

/**
 * Type for a straight tile.
 * @type {string}
 */
export const STRAIGHT_TILE_TYPE = 'straight-tile';

/**
 * Type for a curved tile.
 * @type {string}
 */
export const CURVED_TILE_TYPE = 'curved-tile';

/**
 * Type for an enlarged curve tile.
 * @type {string}
 */
export const CURVED_TILE_ENLARGED_TYPE = 'curved-tile-enlarged';

/**
 * @type {string[]} - The list of allowed tile directions.
 * @private
 */
const allowedDirection = [TILE_DIRECTION_RIGHT, TILE_DIRECTION_LEFT];

/**
 * Checks if a given direction is valid for a tile.
 * @param {*} direction - The direction to check.
 * @returns {boolean} - Returns `true` if the direction is valid.
 */
export const isDirectionValid = direction => allowedDirection.includes(direction);

/**
 * @type {string[]} - The list of allowed tile types.
 * @private
 */
const allowedTypes = [STRAIGHT_TILE_TYPE, CURVED_TILE_TYPE, CURVED_TILE_ENLARGED_TYPE];

/**
 * Checks if a given type is valid for a tile.
 * @param {*} type - The type to check.
 * @returns {boolean} - Returns `true` if the type is valid.
 */
export const isTypeValid = type => allowedTypes.includes(type);

/**
 * Flips the direction of a tile.
 * @param {string} direction - The original direction
 * @returns {string} - The flipped direction.
 */
export const flipTileDirection = direction => {
    if (direction === TILE_DIRECTION_LEFT) {
        return TILE_DIRECTION_RIGHT;
    }
    return TILE_DIRECTION_LEFT;
};
