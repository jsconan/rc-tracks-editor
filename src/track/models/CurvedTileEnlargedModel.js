/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { CURVED_TILE_ENLARGED_TYPE, TILE_DIRECTION_LEFT } from '../helpers';
import { TileModel } from './TileModel.js';
import { Vector2D } from '../../core/models';

/**
 * Represents a track tile for an enlarged curve.
 */
export class CurvedTileEnlargedModel extends TileModel {
    /**
     * Sets the size factor relative to a track section.
     * 1 means the tile fits 1 tile section in each direction.
     * @param {number} ratio - The size factor relative to a track section.
     * @returns {CurvedTileEnlargedModel} - Chains the instance.
     */
    setRatio(ratio) {
        this.ratio = Math.max(1, Math.floor(Math.abs(ratio)));
        return this;
    }

    /**
     * Computes the angle for rotating the tile to the expected direction.
     * @returns {number}
     */
    getDirectionAngle() {
        if (this.direction === TILE_DIRECTION_LEFT) {
            return Vector2D.RIGHT_ANGLE;
        }

        return 0;
    }

    /**
     * Computes the angle of the curve with respect to the ratio.
     * @returns {number}
     */
    getCurveAngle() {
        return Vector2D.RIGHT_ANGLE;
    }

    /**
     * Computes the length of the curve side if any.
     * @returns {number}
     */
    getCurveSide() {
        return this.length / 2;
    }

    /**
     * Computes the number of barrier chunks on each side with respect to the ratio.
     * @returns {number}
     */
    getSideBarrierChunks() {
        return (this.specs.barrierChunks * this.ratio) / 2;
    }

    /**
     * Computes the number of barrier chunks for the inner curve with respect to the ratio.
     * @returns {number}
     */
    getInnerBarrierChunks() {
        if (this.ratio <= 1) {
            return this.specs.barrierChunks / 2;
        }
        return this.specs.barrierChunks * this.ratio;
    }

    /**
     * Computes the number of barrier chunks for an outer curve with respect to the ratio.
     * @returns {number}
     */
    getOuterBarrierChunks() {
        return (this.specs.barrierChunks / 2) * this.ratio;
    }

    /**
     * Computes the coordinates of the center point.
     * @param {number} x - The X-coordinate of the tile.
     * @param {number} y - The Y-coordinate of the tile.
     * @param {number} angle - The rotation angle of the tile.
     * @returns {Vector2D}
     */
    getCenterCoord(x = 0, y = 0, angle = 0) {
        const start = new Vector2D(x, y);
        const center = start.addScalarY(this.specs.length * (this.ratio - 0.5));

        return center.rotateAround(angle, start);
    }

    /**
     * Computes the coordinates of the output point with respect to the tile direction.
     * @param {number} x - The X-coordinate of the tile.
     * @param {number} y - The Y-coordinate of the tile.
     * @param {number} angle - The rotation angle of the tile.
     * @returns {Vector2D}
     */
    getOutputCoord(x = 0, y = 0, angle = 0) {
        const start = new Vector2D(x, y);

        const radius = this.getInnerRadius() + this.specs.width / 2;
        const curveAngle = this.getCurveAngle();

        let center;
        if (this.direction === TILE_DIRECTION_LEFT) {
            center = start.addScalarX(radius);
        } else {
            center = start.subScalarX(radius);
        }

        return Vector2D.polar(radius, curveAngle, center).rotateAround(angle, start);
    }

    /**
     * Computes the angle of the output point with respect to the tile direction.
     * @param {number} angle - The rotation angle of the tile.
     * @returns {number}
     */
    getOutputAngle(angle = 0) {
        if (this.direction === TILE_DIRECTION_LEFT) {
            return Vector2D.degrees(angle - Vector2D.RIGHT_ANGLE);
        }

        return Vector2D.degrees(angle + Vector2D.RIGHT_ANGLE);
    }

    /**
     * Computes the coordinates of the edge point with respect to the tile direction.
     * @param {number} x - The X-coordinate of the tile.
     * @param {number} y - The Y-coordinate of the tile.
     * @param {number} angle - The rotation angle of the tile.
     * @returns {Vector2D}
     */
    getEdgeCoord(x = 0, y = 0, angle = 0) {
        const start = new Vector2D(x, y);

        const center = start.addScalarY(this.specs.length * (this.ratio - 0.5));
        const radius = this.getOuterRadius();
        let curveAngle;

        if (this.direction === TILE_DIRECTION_LEFT) {
            curveAngle = Vector2D.RIGHT_ANGLE + this.getCurveAngle() / 2;
        } else {
            curveAngle = Vector2D.RIGHT_ANGLE - this.getCurveAngle() / 2;
        }

        return Vector2D.polar(radius, curveAngle, center).rotateAround(angle, start);
    }
}

/**
 * The type of tile.
 * @constant {string} CurvedTileEnlargedModel.TYPE
 */
Object.defineProperty(CurvedTileEnlargedModel, 'TYPE', {
    value: CURVED_TILE_ENLARGED_TYPE,
    writable: false,
    enumerable: true,
    configurable: true
});
