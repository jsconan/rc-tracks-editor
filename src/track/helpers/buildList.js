/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { validateAPI } from '../../core/helpers';
import { Vector2D } from '../../core/models';
import { TILE_DIRECTION_LEFT } from '../../tile/helpers';

/**
 * Process a list of tiles for rendering, computing the coordinates of each tile.
 * @param {*} list - The list of tiles from which build the track. It must implement a map method.
 * @param {object} config - A set of config options.
 * @param {number} [config.startX] - The X-coordinate of the first tile.
 * @param {number} [config.startY] - The Y-coordinate of the first tile.
 * @param {number} [config.tileAngle] - The rotation angle of each tile.
 * @param {number} [config.tileWidth] - The reserved width for each tile.
 * @param {number} [config.tileHeight] - The reserved height for each tile.
 * @param {number} [config.hPadding] - An horizontal padding added around the tiles.
 * @param {number} [config.vPadding] - A vertical padding added around the tiles.
 * @param {boolean} [config.centered] - Whether or not center the tiles.
 * @param {boolean} [config.aligned] - Whether or not align the tiles.
 * @param {boolean} [config.vertical] - Position the tiles vertically.
 * @returns {trackCoord} - Returns the list of coordinates.
 * @throws {TypeError} - If the given list does not implement the map method.
 */
export default (
    list,
    {
        startX = 0,
        startY = 0,
        tileAngle = 0,
        tileWidth = 0,
        tileHeight = 0,
        hPadding = 0,
        vPadding = 0,
        centered = false,
        aligned = false,
        vertical = false
    } = {}
) => {
    validateAPI(list, ['map']);

    const topLeft = new Vector2D();
    const bottomRight = new Vector2D();
    let tileX = startX;
    let tileY = startY;

    const tiles = list.map(model => {
        const curveAngle = model.getCurveAngle();
        const { id, type, direction, ratio } = model;
        const alignedAngle = curveAngle / 2;
        const directionAngle =
            direction === TILE_DIRECTION_LEFT ? Vector2D.STRAIGHT_ANGLE * 2 - alignedAngle : alignedAngle;
        const angle = tileAngle - (aligned && curveAngle < Vector2D.RIGHT_ANGLE ? directionAngle : 0);
        const rect = model.getBoundingRect(0, 0, angle);
        const width = Math.max(tileWidth, rect.width) + hPadding * 2;
        const height = Math.max(tileHeight, rect.height) + vPadding * 2;
        const dx = centered ? Math.max(0, tileWidth - rect.width) / 2 : 0;
        const dy = centered ? Math.max(0, tileHeight - rect.height) / 2 : 0;
        const topX = tileX - (centered && vertical ? width / 2 : 0) - hPadding;
        const topY = tileY - (!centered || vertical ? 0 : height / 2) - vPadding;
        const x = topX - rect.x + hPadding + dx;
        const y = topY - rect.y + vPadding + dy;

        topLeft.x = Math.min(topLeft.x, topX);
        topLeft.y = Math.min(topLeft.y, topY);

        bottomRight.x = Math.max(bottomRight.x, topX + width);
        bottomRight.y = Math.max(bottomRight.y, topY + height);

        if (vertical) {
            tileY += height;
        } else {
            tileX += width;
        }

        return { id, type, direction, ratio, x, y, angle, rect, model };
    });

    const { x, y } = topLeft;
    const { x: width, y: height } = bottomRight.sub(topLeft);

    return { x, y, width, height, tiles };
};

/**
 * @typedef {import('../models/TrackBuilder.js').trackCoord} trackCoord
 */
