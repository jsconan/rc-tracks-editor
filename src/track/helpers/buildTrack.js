/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { validateAPI } from '../../core/helpers';
import { Vector2D } from '../../core/models';

/**
 * Process a list of tiles for rendering a track, computing the coordinates of each tile.
 * @param {*} list - The list of tiles from which build the track. It must implement a map method.
 * @param {object} config - A set of config options.
 * @param {number} [config.startX] - The X-coordinate of the first tile.
 * @param {number} [config.startY] - The Y-coordinate of the first tile.
 * @param {number} [config.startAngle] - The rotation angle of the first tile.
 * @param {number} [config.hPadding] - An horizontal padding added around the track.
 * @param {number} [config.vPadding] - A vertical padding added around the tracks.
 * @returns {trackCoord} - Returns the list of coordinates.
 * @throws {TypeError} - If the given list does not implement the map method.
 */
export default (list, { startX = 0, startY = 0, startAngle = 0, hPadding = 0, vPadding = 0 } = {}) => {
    validateAPI(list, ['map']);

    const topLeft = new Vector2D();
    const bottomRight = new Vector2D();
    let inputX = startX;
    let inputY = startY;
    let inputAngle = startAngle;

    const tiles = list.map(model => {
        const rect = model.getBoundingRect(inputX, inputY, inputAngle);
        const { x, y, angle } = rect.input;
        const { id, type, direction, ratio } = model;

        inputX = rect.output.x;
        inputY = rect.output.y;
        inputAngle = rect.output.angle;

        topLeft.x = Math.min(topLeft.x, rect.x);
        topLeft.y = Math.min(topLeft.y, rect.y);

        bottomRight.x = Math.max(bottomRight.x, rect.x + rect.width);
        bottomRight.y = Math.max(bottomRight.y, rect.y + rect.height);

        return { id, type, direction, ratio, x, y, angle, rect, model };
    });

    const { x, y } = topLeft.subCoord(hPadding, vPadding);
    const { x: width, y: height } = bottomRight.sub(topLeft).addCoord(hPadding * 2, vPadding * 2);

    return { x, y, width, height, tiles };
};

/**
 * @typedef {object} trackCoord - Represents a track ready to be rendered.
 * @property {number} x - The left coordinate of the track.
 * @property {number} y - The top coordinate of the track.
 * @property {number} width - The width of the track.
 * @property {number} height - The height of the track.
 * @property {tileCoord[]} tiles - The list of tiles.
 */

/**
 * @typedef {object} tileCoord - Represents a positioned tile.
 * @property {string} id - The unique identifier of the tile.
 * @property {string} type - The type of tile.
 * @property {string} direction - The direction of the tile.
 * @property {number} ratio - The size ratio of the tile.
 * @property {number} x - The left coordinate of the tile.
 * @property {number} y - The top coordinate of the tile.
 * @property {number} angle - The rotation angle of the tile.
 * @property {TileModel} model - A reference to the tile model.
 */

/**
 * @typedef {import('../../tile/models').TileModel} TileModel
 */
