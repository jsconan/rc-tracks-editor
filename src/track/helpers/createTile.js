/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import {
    validateType,
    CURVED_TILE_ENLARGED_TYPE,
    CURVED_TILE_TYPE,
    STRAIGHT_TILE_TYPE,
    TILE_DIRECTION_RIGHT
} from '../../tile/helpers';
import { CurvedTileEnlargedModel, CurvedTileModel, StraightTileModel } from '../../tile/models';

/**
 * @type {object} - Maps the types of tile to their respective model.
 * @private
 */
const modelsMap = {
    [STRAIGHT_TILE_TYPE]: StraightTileModel,
    [CURVED_TILE_TYPE]: CurvedTileModel,
    [CURVED_TILE_ENLARGED_TYPE]: CurvedTileEnlargedModel
};

/**
 * Creates the tile model with respect to the given type.
 * @param {TileSpecifications} specs - The specifications for the tiles.
 * @param {string} type - The type of tile.
 * @param {string} direction - The direction of the tile.
 * @param {number} ratio - The size ratio of the tile.
 * @returns {TileModel} - Returns a tile model of the expected type.
 * @throws {TypeError} - If the given specifications object is not valid.
 * @throws {TypeError} - If the given type is not valid.
 * @throws {TypeError} - If the given direction is not valid.
 * @private
 */
export default (specs, type = STRAIGHT_TILE_TYPE, direction = TILE_DIRECTION_RIGHT, ratio = 1) => {
    validateType(type);
    return new modelsMap[type](specs, direction, ratio);
};
