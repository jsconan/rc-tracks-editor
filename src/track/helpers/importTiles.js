/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { TileList } from '../../tile/models';
import createTile from './createTile.js';

/**
 * Imports a list of tiles from exported data.
 * @param {TileList} list - The list in which import the tiles.
 * @param {tileExport[]} data - An iterable object containing exported tiles.
 * @returns {TileList} - Returns the list.
 */
export default (list, data) => {
    TileList.validateInstance(list);

    if (!data || !data[Symbol.iterator]) {
        return list;
    }

    const dataIterator = data[Symbol.iterator]();
    const loadIterator = {
        next() {
            const next = dataIterator.next();

            if (next.done) {
                return next;
            }

            const { type, direction, ratio } = next.value || {};
            const tile = createTile(list.specs, type, direction, ratio);

            return { done: false, value: tile };
        },

        [Symbol.iterator]() {
            return this;
        }
    };

    list.load(loadIterator);

    return list;
};

/**
 * @typedef {import('../../tile/models/TileModel.js').tileExport} tileExport
 */
