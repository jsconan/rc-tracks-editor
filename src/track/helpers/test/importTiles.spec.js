/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { CURVED_TILE_TYPE, TILE_DIRECTION_LEFT, TILE_DIRECTION_RIGHT } from '../../../tile/helpers';
import { TileSpecifications } from '../../../tile/config';
import { TileList } from '../../../tile/models';
import importTiles from '../importTiles.js';

const laneWidth = 80;
const barrierWidth = 5;
const barrierChunks = 4;
const specs = new TileSpecifications(laneWidth, barrierWidth, barrierChunks);

describe('importTiles', () => {
    it('is a function', () => {
        expect(importTiles).toEqual(expect.any(Function));
    });

    describe('throws error', () => {
        it('if the given list is not a TileList', () => {
            expect(() => importTiles({})).toThrow('The object must be an instance of TileList!');
        });

        it('when trying to import a tile with an invalid type', () => {
            const list = new TileList(specs);
            expect(() => importTiles(list, [{ type: '' }])).toThrow('A valid type of tile is needed!');
        });

        it('when trying to import a tile with an invalid direction', () => {
            const list = new TileList(specs);
            expect(() => importTiles(list, [{ type: CURVED_TILE_TYPE, direction: '' }])).toThrow(
                'A valid direction is needed!'
            );
        });
    });

    it('imports tiles from an object', () => {
        const list = new TileList(specs);
        const data = [
            {
                type: CURVED_TILE_TYPE,
                direction: TILE_DIRECTION_LEFT,
                ratio: 1
            },
            null,
            {
                type: CURVED_TILE_TYPE,
                direction: TILE_DIRECTION_RIGHT,
                ratio: 1
            }
        ];

        const callback = jest.fn();
        list.on('load', callback);

        expect(importTiles(list, {})).toBe(list);
        expect(list).toMatchSnapshot();
        expect(callback).toHaveBeenCalledTimes(0);

        expect(importTiles(list, data)).toBe(list);
        expect(list).toMatchSnapshot();
        expect(callback).toHaveBeenCalledTimes(1);
    });
});
