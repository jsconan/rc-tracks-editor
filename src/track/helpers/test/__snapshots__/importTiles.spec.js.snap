// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`importTiles imports tiles from an object 1`] = `
TileList {
  "delegate": [Function],
  "emit": [Function],
  "list": Array [],
  "listens": [Function],
  "off": [Function],
  "on": [Function],
  "once": [Function],
  "specs": TileSpecifications {
    "barrierChunks": 4,
    "barrierWidth": 5,
    "laneWidth": 80,
    "maxRatio": 4,
  },
  "subscribe": [Function],
}
`;

exports[`importTiles imports tiles from an object 2`] = `
TileList {
  "delegate": [Function],
  "emit": [Function],
  "list": Array [
    CurvedTileModel {
      "direction": "left",
      "id": "id-0",
      "ratio": 1,
      "specs": TileSpecifications {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "maxRatio": 4,
      },
    },
    StraightTileModel {
      "direction": "right",
      "id": "id-1",
      "ratio": 1,
      "specs": TileSpecifications {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "maxRatio": 4,
      },
    },
    CurvedTileModel {
      "direction": "right",
      "id": "id-2",
      "ratio": 1,
      "specs": TileSpecifications {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "maxRatio": 4,
      },
    },
  ],
  "listens": [Function],
  "off": [Function],
  "on": [Function],
  "once": [Function],
  "specs": TileSpecifications {
    "barrierChunks": 4,
    "barrierWidth": 5,
    "laneWidth": 80,
    "maxRatio": 4,
  },
  "subscribe": [Function],
}
`;
